File: lekcje/04_drzewo_binarne/przyklady/dzrzewo_binarne_python/drzewo_binarne.py
class Wezel:
    def __init__(self, klucz):
        self.klucz = klucz
        self.lewe = None
        self.prawe = None


class DrzewoBinarne:
    def __init__(self):
        self.korzen = None

    def dodaj(self, klucz):
        if self.korzen is None:
            self.korzen = Wezel(klucz)
        else:
            self._dodaj(klucz, self.korzen)

    def _dodaj(self, klucz, wezel):
        if klucz < wezel.klucz:
            if wezel.lewe is None:
                wezel.lewe = Wezel(klucz)
            else:
                self._dodaj(klucz, wezel.lewe)
        else:
            if wezel.prawe is None:
                wezel.prawe = Wezel(klucz)
            else:
                self._dodaj(klucz, wezel.prawe)

    def wyswietl(self):
        if self.korzen is not None:
            self._wyswietl(self.korzen)

    def _wyswietl(self, wezel):
        if wezel is not None:
            self._wyswietl(wezel.lewe)
            print(str(wezel.klucz) + " ", end="")
            self._wyswietl(wezel.prawe)


# Przyklad uzycia
if __name__ == "__main__":
    drzewo = DrzewoBinarne()
    drzewo.dodaj(10)
    drzewo.dodaj(5)
    drzewo.dodaj(15)
    drzewo.dodaj(2)
    drzewo.dodaj(7)

    print("Elementy drzewa binarnego (w porzadku rosnacym):")
    drzewo.wyswietl()

File: lekcje/04_drzewo_binarne/przyklady/dzrzewo_binarne_c/drzewo_binarne.c
#include "drzewo_binarne.h"
#include <stdio.h>
#include <stdlib.h>

// Funkcja tworzaca nowy wezel drzewa
Wezel *utworz_wezel(int klucz) {
  Wezel *nowy_wezel = (Wezel *)malloc(sizeof(Wezel));
  nowy_wezel->klucz = klucz;
  nowy_wezel->lewe = NULL;
  nowy_wezel->prawe = NULL;
  return nowy_wezel;
}

// Funkcja dodajaca nowy wezel do drzewa
Wezel *dodaj_wezel(Wezel *korzen, int klucz) {
  if (korzen == NULL) {
    return utworz_wezel(klucz);
  }

  if (klucz < korzen->klucz) {
    korzen->lewe = dodaj_wezel(korzen->lewe, klucz);
  } else if (klucz > korzen->klucz) {
    korzen->prawe = dodaj_wezel(korzen->prawe, klucz);
  }

  return korzen;
}

// Funkcja wyswietlajaca drzewo binarne (przechodzenie inorder)
void wyswietl_drzewo(Wezel *korzen) {
  if (korzen != NULL) {
    wyswietl_drzewo(korzen->lewe);
    printf("%d ", korzen->klucz);
    wyswietl_drzewo(korzen->prawe);
  }
}
